README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# stressaddition

This R package makes it possible to model tri-phasic
concentration-response relationships using the stress addition approach.
It is useful for the analysis of ecotoxicological data where the
traditional concentration addition or effect addition models are
inadequate. Its main functions are `ecxsys()` and `multi_tox()`.

`ecxsys()` implements **EC<sub>x-SyS</sub>**, the tri-phasic
concentration-response model introduced in [Liess, M., Henz, S. &
Knillmann, S. Predicting low-concentration effects of pesticides. Sci
Rep 9, 15248 (2019)](https://doi.org/10.1038/s41598-019-51645-4). It is
applicable to modelling ecotoxicological experiments with and without
environmental stress where the response contains a hormesis effect.

`multi_tox()` implements **Multi-TOX**, a model for binary mixtures of
toxicants where each toxicant exhibits a tri-phasic
concentration-response relationship. See [Liess, M., Henz, S. & Shahid,
N. Modeling the synergistic effects of toxicant mixtures. Environ Sci
Eur 32, 119 (2020)](https://doi.org/10.1186/s12302-020-00394-7).

The EC<sub>x-SyS</sub> and Multi-TOX models are also available as part
of the [Indicate app](http://www.systemecology.eu/indicate) which offers
an easy to use graphical user interface.

## Installation

This package is available from CRAN:

``` r
install.packages("stressaddition")
```

You can also get the development version (potentially unstable) from
GitLab:

``` r
install.packages("remotes")
remotes::install_gitlab("oekotox/stressaddition", host = "git.ufz.de")
```

Alternatively, there are binary and source builds of the current release
and older versions available for download from the [releases
page](https://git.ufz.de/oekotox/stressaddition/-/releases).

## Examples

### EC<sub>x-SyS</sub>

Model a concentration-response relationship with hormesis:

``` r
library(stressaddition)
model_a <- ecxsys(
    concentration = c(0, 0.05, 0.5, 5, 30),
    survival_tox_observed = c(90, 81, 92, 28, 0),
    survival_tox_env_observed = c(29, 27, 33, 5, 0),
    hormesis_concentration = 0.5
)
```

Calculate the LC<sub>50</sub> under the influence of toxicant and system
stress and the LC<sub>10</sub> under the influence of toxicant,
environmental and system stress:

``` r
lc(model_a, "survival_tox_sys", 50)
#> $response
#> [1] 44.95368
#> 
#> $concentration
#> [1] 3.375735
lc(model_a, "survival_tox_env_sys", 10)
#> $response
#> [1] 26.41904
#> 
#> $concentration
#> [1] 0.0008571244
```

Plot the survival and the system stresses:

``` r
par(mfrow = c(2, 1), mar = c(4, 4, 0.5, 0.1))
plot_survival(model_a)
plot_stress(model_a)
```

![](images/README-example_plot-1.png)<!-- -->

### Multi-TOX

Define an additional single toxicant model and calculate the survival
for some binary concentration mixtures:

``` r
model_b <- ecxsys(
    concentration = c(0, 0.01, 0.1, 1, 10, 100),
    survival_tox_observed = c(96, 89, 91, 57, 9, 0),
    hormesis_concentration = 0.1
)
multi_tox(
    model_a,
    model_b,
    concentration_a = c(0.1, 0.3, 2, 15),
    concentration_b = c(0.04, 0.1, 1, 13)
)[, 1:3]
#>   concentration_a concentration_b survival
#> 1             0.1            0.04 84.44956
#> 2             0.3            0.10 73.53734
#> 3             2.0            1.00 13.38661
#> 4            15.0           13.00  0.00000
```

## Citation

Please cite this package if you use it in your analysis. See
`citation("stressaddition")` for details.

## Copyright and License

Copyright (c) 2020,  
Helmholtz-Zentrum fuer Umweltforschung GmbH - UFZ.  
All rights reserved.

The code is a property of:

Helmholtz-Zentrum fuer Umweltforschung GmbH - UFZ  
Registered Office: Leipzig  
Registration Office: Amtsgericht Leipzig  
Trade Register: Nr. B 4703  
Chairman of the Supervisory Board: MinDirig’in Oda Keppler  
Scientific Director: Prof. Dr. Georg Teutsch  
Administrative Director: Dr. Sabine König

stressaddition is free software: you can redistribute it and/or modify  
it under the terms of the GNU General Public License as published by  
the Free Software Foundation, either version 3 of the License, or  
(at your option) any later version.

This program is distributed in the hope that it will be useful,  
but WITHOUT ANY WARRANTY; without even the implied warranty of  
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the  
GNU General Public License for more details.

You should have received a copy of the GNU General Public License  
along with this program. If not, see <https://www.gnu.org/licenses/>.


NEWS.md
# stressaddition 3.1.0

* Improved the documentation of the `which` argument in `plot_survival()` and `plot_stress()`.
* Changed the default value of the `which` argument in the plot functions. Now it contains the proper default curve names. If it is `NA` only the axes and labels will get drawn.


# stressaddition 3.0.3

* Added references to the journal article about the Multi-Tox model which was recently published.


# stressaddition 3.0.2

* Fixed a bug where the plotting functions printed `NULL` to the console.
* Changed maintainer e-mail address to ensure long-term maintainability.
* Internal improvements related to package testing.


# stressaddition 3.0.1

* Added missing return values in documentation to comply with CRAN repository policy.
* Added a link to the paper in the description file.


# stressaddition 3.0.0

## Breaking changes
* Renamed all instances of "effect" to "survival".
* Renamed all instances of "ec" to "lc".
* Renamed `predict_mixture()`, which was a temporary development name, to `multi_tox()`.
* The argument `proportion_ca` in the mixture model `multi_tox()` was renamed and its value reversed. It is now called `sa_contribution` and specifies the proportion of stress addition in the calculation of toxicant stress. To convert your code from the old version use this equation: `sa_contribution = 1 - proportion_ca`.
* Renamed `stress_tox_sam` to `stress_tox_sa` in the output of `multi_tox()`.

## Bugfixes
* Fixed a bug where `plot_stress()` with argument `which = NULL` would result in an error. Now it correctly draws the axes without data.
* Fixed some errors in the documentation and improved the examples.

## New
* Exported function `log10_ticks()` for calculating tick mark labels and positions on a base 10 logarithmic axis.
* Added example data set `multiple_stress` for use with `multi_tox()`.
* Various minor changes to prepare for CRAN submission.


# stressaddition 2.7.0

* Fixed some spelling mistakes.
* `predict_mixture()` now also returns the various stresses.


# stressaddition 2.6.0

* The `curves` data frame in the output of `ecxsys()` now contains a column with the concentrations which are used for the plot functions in this package. This is useful for generating a nicer concentration axis.
* Changes to `ec()`:
    * Renamed `response_value` to `effect` in the output list.
    * `response_level` of 0 or 100 is now allowed. 0 returns the concentration 0 and 100 returns the concentration `Inf`. Previously this resulted in an error.
    * It is now possible to set the reference to a custom value, for example 100.


# stressaddition 2.5.0

* Fixed unintended behaviour in `plot_effect()` and `plot_stress()` where supplying an empty vector caused the four standard curves to show. Now setting `which` to an empty vector or `NULL` shows just the axes. The default value is NA.
* Renamed the `mixture_effect` column in the `predict_mixture` output data frame to `effect`.
* Remove the restriction that the concentration vectors in `predict_mixture()` must be the same length. The longer length must be a multiple of the shorter length because the shorter vector gets recycled to the longer length.


# stressaddition 2.4.0

* Improved `plot_effect()` and `plot_stress()`. You can now control whether the observed values (the points) should be plotted using the `which` argument.
* Renamed `sys_tox_not_fitted` and `sys_tox_env_not_fitted` to `sys_tox_observed` and `sys_tox_env_observed` in the output of `ecxsys()`.


# stressaddition 2.3.0

* `predict_mixture()` now accepts multiple values for the concentration of the second toxicant. Both concentration vectors must be the same length.
* `predict_mixture()` now returns a data frame with the concentrations and effects. Previously it was only a vector of effects.
* `predict_mixture()` received a new argument "effect_max" which scales the returned effect values.
* Renamed the arguments of `predict_mixture()` to use underscore letters a and b instad of 1 and 2. For example model_1 is now model_a.


# stressaddition 2.2.1

* Improved documentation of `predict_mixture()` and included example of symmetry.


# stressaddition 2.2.0

* `ec()` now raises an error if the curve does not cross the desired response level.
* `ecxsys()` gained a new argument `curves_concentration_max` which allows setting the maximum concentration of the predicted curves.


# stressaddition 2.1.1

* Restore the default behaviour of `plot_effect()` to also show `effect_tox` and `effect_tox_env`.


# stressaddition 2.1.0

* The functions `plot_effect()` and `plot_stress()` gained a `which` argument that controls which curves are plotted. Consequently, the `show_LL5_model` argument of `plot_effect()` was removed.
* Added arguments `xlab` and `ylab` to `plot_stress`.
* Added argument `main` to both plot functions.
* Changed some colors of the stress curves so they better match with the colors of related effect curves.
* Added `predict_mixture()` for the prediction of the effects of mixtures of two toxicants.
* Fixed documentation of `ecxsys()` and `predict_ecxsys()`.


# stressaddition 2.0.0

* Changed the order of arguments in `ecxsys()`.
* Removed `hormesis_index` argument from `ecxsys()`. Use `hormesis_concentration` instead.
* New function `predict_ecxsys()` replaces `fn()` from the `ecxsys()` output.
* Renamed the arguments in `ec()`.
* Made `ec()` more flexible. It now also accepts a data.frame with a concentration column and a column of response values.
* Added LL5 curves to the legend of `plot_effect()`.
* Replaced every occurrence of "simple" in variable names with "LL5".
* Replaced every occurrence of "sys_stress" in variable names with "sys" because the extra "stress" was redundant.
* Renamed `plot_system_stress()` to `plot_stress()` because it is planned to plot more stresses with this function in a future update.
* Changed the order of the columns in the output of `predict_ecxsys()`.
* Improved the internal structure of the package.

R/plot_ecxsys.R
# Copyright (C) 2020  Helmholtz-Zentrum fuer Umweltforschung GmbH - UFZ
# See file inst/COPYRIGHTS for details.
#
# This file is part of the R package stressaddition.
#
# stressaddition is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


# This is the shared documentation for the plotting functions.


#' Plot the results of the ECx-SyS model
#'
#' Plot the observed and modeled survivals and stresses.
#'
#' @name plot_ecxsys
#'
#' @param model The model object returned from \code{\link{ecxsys}}.
#' @param which A vector of curve names to plot. Allowed are all values of
#'   \code{names(model$curves)} except \code{"concentration"}. See the
#'   \emph{Value} section of \code{\link{predict_ecxsys}} for descriptions of
#'   these names. Additionally, you can use \code{"survival_tox_observed"} and
#'   \code{"survival_tox_env_observed"} for the observed survival and
#'   \code{"sys_tox_observed"} and \code{"sys_tox_env_observed"} for the
#'   observed Sys. Use \code{"all"} to plot everything or \code{NA} to only draw
#'   the axis and labels. Invalid names will be silently ignored.
#'
#' @param show_legend Should the plot include a legend? Defaults to \code{FALSE}
#'   because it may cover some parts of the plot depending on the plot size and
#'   the number of elements shown.
#' @param xlab,ylab,main Axis labels and title.
#'
#' @return None
#'
#' @examples model <- ecxsys(
#'     concentration = c(0, 0.05, 0.5, 5, 30),
#'     hormesis_concentration = 0.5,
#'     survival_tox_observed = c(90, 81, 92, 28, 0),
#'     survival_tox_env_observed = c(29, 27, 33, 5, 0)
#' )
#' plot_survival(model, show_legend = TRUE)
#' plot_stress(model, show_legend = TRUE)
#'
#' # Plot all curves:
#' plot_survival(model, which = "all")
#' plot_stress(model, which = "all")
#'
#' # Plot only some selected curves:
#' plot_survival(model, which = c("survival_tox_sys", "survival_tox_env_sys"))
#' plot_stress(model, which = c("sys_tox", "sys_tox_env"))
#'
#' # Plot only the observed values:
#' plot_survival(model, which = c("survival_tox_observed", "survival_tox_env_observed"))
#' plot_stress(model, which = c("sys_tox_observed", "sys_tox_env_observed"))
NULL
